source/interface/azure_iot_transport_interface.h

```diff
@@ -0,0 +1,182 @@
+ * <br>
+ * This function is expected to populate a buffer, with bytes received from the
+ * transport, and return the number of bytes placed in the buffer.
+ * In the case of TLS over TCP, @ref AzureIoTTransportRecv_t is typically implemented by
+ * calling the TLS layer function to receive data. In case of plaintext TCP
+ * without TLS, it is typically implemented by calling the TCP layer receive
+ * function. @ref AzureIoTTransportRecv_t may be invoked multiple times by the protocol
+ * library, if fewer bytes than were requested to receive are returned.
+ * <br><br>
+ * <b>Example code:</b>
+ * @code{c}
+ * int32_t myNetworkRecvImplementation( struct NetworkContext * pNetworkContext,
+ *                                      void * pBuffer,
+ *                                      size_t bytesToRecv )
+ * {
+ *     int32_t bytesReceived = 0;
+ *     bytesReceived = TLSRecv( pNetworkContext->tlsContext,
+ *                              pBuffer,
+ *                              bytesToRecv,
+ *                              MY_SOCKET_TIMEOUT );
+ *     if( bytesReceived < 0 )
+ *     {
+ *         // Handle socket error.
+ *     }
+ *     // Handle other cases.
+ *
+ *     return bytesReceived;
+ * }
+ * @endcode
+ * <br>
+ * -# Implementing @ref AzureIoTTransportSend_t<br><br>
+ * @snippet this define_transportsend
+ * <br>
+ * This function is expected to send the bytes, in the given buffer over the
+ * transport, and return the number of bytes sent.
+ * In the case of TLS over TCP, @ref AzureIoTTransportSend_t is typically implemented by
+ * calling the TLS layer function to send data. In case of plaintext TCP
+ * without TLS, it is typically implemented by calling the TCP layer send
+ * function. @ref AzureIoTTransportSend_t may be invoked multiple times by the protocol
+ * library, if fewer bytes than were requested to send are returned.
+ * <br><br>
+ * <b>Example code:</b>
+ * @code{c}
+ * int32_t myNetworkSendImplementation( struct NetworkContext * pNetworkContext,
+ *                                      const void * pBuffer,
+ *                                      size_t bytesToSend )
+ * {
+ *     int32_t bytesSent = 0;
+ *     bytesSent = TLSSend( pNetworkContext->tlsContext,
+ *                          pBuffer,
+ *                          bytesToSend,
+ *                          MY_SOCKET_TIMEOUT );
+ *     if( bytesSent < 0 )
+ *     {
+ *         // Handle socket error.
+ *     }
+ *     // Handle other cases.
+ *
+ *     return bytesSent;
+ * }
+ * @endcode
+ */
+
+/**
+ * @transportstruct
+ * @brief The NetworkContext is an incomplete type. An implementation of this
+ * interface must define struct NetworkContext for the system requirements.
+ * This context is passed into the network interface functions.
+ */
+/* @[define_networkcontext] */
+struct NetworkContext;
+/* @[define_networkcontext] */
+
+/**
+ * @transportcallback
+ * @brief Transport interface for receiving data on the network.
```

Comment:
Nit - consider making it more clear it's a user thing, such as "User defined function for receiving data on the network."

I'm not sure if using words like `interface` is for the function name itself.  Interface isn't strictly a C concept in 1st place or when collequially i've seen it used it's kind of like C SDK `xio` (so it's a structure with function pointers, but not an individual function pointer).

Reply:
Updated

---
In reply to: [598970825](https://github.com/Azure/azure-iot-middleware-freertos/pull/44#discussion_r598970825) [](ancestors = 598970825)